
\chapter[Algorithms for Analysis of Multi-Patient scRNA-seq Experiments][Algorithms for Analysis of Multi-Patient scRNA-seq Experiments]{Algorithms for Analysis of Multi-Patient scRNA-seq Experiments}

\section{Introduction}

The previous chapter describes an effort to generate a deep transcriptional map of immune cell states in human breast cancer.
Using fluorescence-assisted cell-sorting and single-cell sequencing, we constructed an atlas of the tumor immune ecosystem microenvironment comprising 47,016 CD45\textsuperscript{+} cells collected from 8 primary breast carcinomas from treatment naive patients. 
The extracted tumors had multiple types, including estrogen receptor (ER\textsuperscript{+}) and progesterone receptor (PR\textsuperscript{+}) positive, human epidermal growth factor receptor 2 amplified (Her2\textsuperscript{+}), and triple negative (TNBC) cancers (Figure~\ref{fig:1b}). 
Through careful modeling of error sources and extensive data filtering, we confirmed that variation in the observed cells stems primarily from biological, and not technical, factors (Figure~\ref{fig:1e}).

At the time, this dataset was the largest that had been generated using InDrop, and the only InDrop dataset generated from multiple human patients.
Our experimental design was substantially more ambitious than the datasets preceding it, which had assayed well characterized model systems such as induced pluripotent stem cell differentiation \citep{Klein2015} or retinal cells \citep{Macosko2015} from mice with identical genetic backgrounds and growth conditions\footnote{Other clinical datasets were generated in parallel which examined data from multiple patients\citep{Tirosh2016}, however we are aware of no study that sequenced samples at equivalent depth.}. 
Our experiment also included multiple genetic backgrounds, tumor types, tissues, and cell types. 
Consequently, the analysis of scRNA-seq data subject to numerous and varying stimuli was a significant and unanswered challenge.  

This chapter describes efforts to address this challenge.
While the last chapter demonstrated approaches to verify the technical quality of the data, this chapter begin by verifying its biological quality by performing some sanity checks on the data using established statistical approaches. 
Specifically, it will analyze each tumor sample independently to confirm that each of the cell types that are expected to be present in TIL isolates are recovered in each patient in the proportions indicated by FACS\@.
The second part of this chapter describes an approach to merge data from different patients through iterative normalization and clustering. 
 
\section{Individual Tumor Samples Capture Complete Human Immune Systems}

To characterize the immune cells extracted from patients, we began by analyzing samples independently to identify their cellular composition and cell type abundances (Figure~\ref{fig:1c},\ref{fig:s1c}).
We reasoned that there would be fewer technical effects that influence cells within a sample than across large numbers of samples. 
Thus, we limited our initial analyses to characterize the cell types within individual tumors. 

To discover cell types in single cell data, a standard approach is to group cells by similarity into "clusters" before comparing the average expression profile of each cluster to a previously identified type.
The best matching type for each cluster can then be considered a good candidate for the type of the cluster. 
There are numerous ways of accomplishing clustering, and there is no clear best approach. 
However, PhenoGraph \citep{Levine2015}, a method that was originally developed to cluster Cytof data, has been adapted to scRNA-seq and appears to have gained community support as the best-practice for clustering cells in a single sample\cite{Shekhar2016,Butler2017}. 

For a distance between two cells to be accurately quantified, cells must first be transformed to approximate ``identically sampled and independently distributed'' data, meaning, practically, that each observed cell expression profile has an equal chance of observing a molecule if it were present in the physical cell. 
This is manifestly untrue for our data, as the vastly different molecule counts achieved for different cells (and different cell types!) obliterate the ``identically sampled'' requirement. % figure here of pre-transformed in-drop data.   

The standard approach to address this problem is median library size normalization\footnote{In the case of droplet-based single-cell data ``normalization'' is a misnomer, as the data, when identically sampled, is better approximated by a negative binomial distribution. However, transformation to the same scale is the important part, rather than the distribution of the data, and so this ``normalization'' method is an appropriate one to this task.}, an approach inspired by bulk RNA-seq approaches \citep{Robinson2010}.
Median library size normalization is a linear scaling approach that sets the total number of molecules in a library to median molecule sum across all observed cells. 
Since the earlier pseudo-bulk analysis of technical replicates showed that the replicates were highly correlated ($\mean{r^2} = 0.96$,Figure~\ref{fig:m1e}), we combined the replicates and normalized them together.  
If each replicate $X_i \in {X_1, X_2, \ldots X_m}$ is composed of $N$ cells, and all replicates measure the same set of $P$ genes, then samples $N_i$ can be scaled according to the median "library size" or median of the total number of molecules in each cell by: 
\[
  m = \median_{\sum_{p=1}^{P}{X_{i,p}}}
  N_i = m * \frac{X_i}{\sum_{p=1}^{P}{X_{i,p}}}
\]
This transformation allows samples which receive different depths of sequencing (different numbers of molecules), to be examined as if they lie on the same scale.
If the observed differences in distribution resulted only from differences in sampling rates between cells, then this transformation is adequate to produce data that are \textit{approximately} identically distributed\footnote{Median library size normalization does not, however, provide a solution for normalizing heterogeneous data from multiple cell types, or enabling comparisons between cells with extremely different sampling rates\citep{Anders2010}}. % add deseq2 reference 

The transformed data, while statistically suitable for distance calculations, contains over 20,000 genes (described computationally as ``features'' or ``dimensions''), which produces a significant computational burden, as many algorithms have scale slowly with increases in features.
However, transcription is controlled through the binding of transcription factors to DNA, each of which control many genes. % citation
As a result, cells responses' to stimuli tend to simultaneously modify or ``co-regulate'' the expression of modules of genes. 

The modularity of transcription implies that high-order correlation structures exist in the gene features of our cells, and as a result, that the data actually lie on relatively low-dimension manifolds within the space of observed features. % segal et al al 2006 A module map showing conditional activity of expression modules in cancer; https://www.ncbi.nlm.nih.gov/pubmed/10591225
As a result, it is possible to reduce the dimensionality of biological data by collapsing it into correlated components without significant loss of information. % need some citations. 

% todo work on this paragraph a bit more. 
To identify a low dimensional representation of the data, we apply Randomized Principal Component Analysis (rPCA) \citep{Halko2009,Rokhlin2009} prior to carrying out subsequent algorithmic steps. % todo Also cite 
The normal PCA method uses singular value decomposition of the data covariance matrix to identify a number of orthogonal components of variation equal to the number of genes in the data. 
A user then examines the fraction of the original variance explained by each component, and selects some number of components $K$, to retain. This is typically done either by selecting components until a certain fraction of the total variation is retained (often 75, 95, or 99\%) or by searching for a ``knee point'', beyond which each components explained variation drops off precipitously. % todo citations, figure. 
Randomized PCA differs accelerates the PCA method bootstrapping over several decompositions of low-rank approximations (with dimension $M$) of the full data covariance matrix, which has dimension $P$, the number of gene features with non-zero observations in at least one cell ($M << P$). \citep{Halko2009}
Randomization can introduce error into low-variation component estimates ($M_i, i > 1000$), however, in scRNA-seq data, most variation is compressed into fewer than 30 components, which rPCA estimates with high accuracy\footnote{In practice, $M$ is set equal to approximately $2K$, the number of components that are expected to be retained. 
For contemporary 3' scRNA-seq data, approximately 25 components are retained, and thus $M$ can be safely set to 50.}.
Thus, use of this algorithm produces a large improvement in speed in what is typically the slowest step of single-sample analysis.  % todo I could introduce an additional figure here comparing the speeds of each method.

A second, and less appreciated, benefit of using PCA to reduce data dimensionality is that it depletes random variation from the data.
By grouping coherent variation into the largest principal components ($k_i, i <= K$), a large proportion of discordant or random variation is pushed into components $k_j, j > K$, which are excluded from analysis. 
Thus, pre-processing with PCA serves both a practical purpose of reducing computation time but also improves data quality.
Because data naturally lie on a low-dimensional manifold, this transformation can be achieved without a significant loss of information. 

Therefore, For each replicate, we applied rPCA to the normalized data $N$ and selected $K$ using the ``knee point'' method \citep{Valle1999}, as described above. 
Because our data was derived from different biological conditions and had different sampling rates, the knee point varied across our samples.
We observed that retention of  $K=6-11$ Principal components per sample produced optimal results, but some iteration over the subsequent clustering and analysis steps was required to determine the correct value. 
The final number of retained PCs in each sample correlated with the pre-normalization library size of the samples, as expected ($r^2 = 0.82$).

The dimension-reduced PCA projection was used as the input to PhenoGraph \citep{Levine2015}, which was applied with default parameters (k=30 nearest neighbors). 
The same principal components were used to generate tSNE projections \citep{Maaten2008}, which were generated with barnes-hut tSNE, implemented in the bhtsne package \href{https://github.com/lvdmaaten/bhtsne}{\emph{https://github.com/lvdmaaten/bhtsne}} (Figure~\ref{fig:1c},\ref{fig:s1c}).
% todo a figure here describing "separation" would be good.

\section{Gross Cell Type Annotation}

Although our immune atlas consists of considerable variability due to the genetic background of the patient, type of tumor, and the tumor microenvironment, it is nevertheless reasonable to expect that high-quality cell profiles should correlate better with cells of their own lineage than those of other immune lineages. 
Therefore, we collected, to our knowledge, all previously generated bulk gene expression profiles of sorted immune cells in humans \citep{Novershtern2011,Jeffrey2006}.
These two studies comprised 37 and 32 microarray experiments taken from sorted normal human immune cells and the same cell populations stimulated by bacterial antigens to provoke immune activation. 

Consistent with our hypothesis that lineage is a stronger source of variance than immune activation or microenvironment, cells of the same lineage clustered together within the microarray experiments regardless of activation state. 
Because the bulk data was generated with microarrays, the data is a complex function of library preparation, but also probe capture efficiency. 
The magnitude of variation in probe capture efficiency is such that microarray analyses are typically limited to making assertions about the relative abundance of genes across samples, rather than measuring the absolute abundance of a gene in a sample. % Significance analysis of microarrays applied to the ionizing radiation response" PNAS 2001 98: 5116-5121, (Apr 24).
As such, in addition to normalizing by library size, microarray data is additionally translated into units of variance by ``Z-scoring'' to remove the effect of abundance on downstream computation. 

This presents a problem for comparison with single-cell sequencing data, as scale is an important predictor of sample quality. As discussed in the previous section, the ambient RNA contamination in single cell experiments means that low-abundance genes are often the result of non-specific RNA diffusion, whereas high-abundance genes are likely specifically expressed in the cells they are detected in. 
As a result, Z-scoring scRNA-seq data prior to analysis significantly degrades the quality of the results. 

To address this problem, genes were stringently filtered such that they must be expressed at an average of at least 1 count in at least one cluster in order to be considered for comparison with the bulk data. % do I need to talk about why we can't use the PCR trick here? 
However, for clusters with low RNA expression or capture, such as naive T-cells, a floor of no fewer than 1000 genes was set to guarantee robust comparisons. 

While a 1-count threshold may sound lenient, it implies that that on average, each cell in a cluster detected the gene. 
Thresholding by cluster was important to avoid biasing the comparisons towards genes that were only present in large cell clusters, which would reduce our power to determine the types of rare populations. 
Both the scRNA-seq and microarray data was then normalized and Z-scored, and the PhenoGraph cluster centroids were correlated with the microarray profiles. 
The correlations were then averaged across lineages and the highest scoring lineage was used to assign each cluster a type. The types are displayed in (Figure~\ref{fig:1c}), and a Heatmap of the correlations is demonstrated in Figure~\ref{fig:2f}.

\begin{figure}
\centering
\includegraphics[width=0.75\textwidth]{Figure1-C.png}
  \caption{t-SNE projection of complete immune systems from two example breast cancer tumors.\ scRNA-seq data for each tumor is processed with pipeline described in chapter 2, library size-normalized, PCA-reduced, and clustered with PhenoGraph. Each dot represents a single-cell colored by its cluster label, and clusters are labeled by cell types, inferred through similarity with bulk profiles. Additional tumors are presented in Figure~\ref{fig:s1b}} 
\label{fig:1c}
\end{figure} 

Our first attempts to annotate clusters with cell types revealed several clusters with low correlations with all bulk immune datasets. 
We reasoned that low correlation was most likely due to one of two possibilities. 
First, low-correlation clusters could be composed of low-quality cells dominated by ribosomal, mitochondrial, or other housekeeping transcripts, which would imply that a cell that should have been filtered out had improperly been retained by SEQC, or it could result because the clusters were not composed of immune cells,.
Alternatively, low-correlation clusters could be composed of stromal or tumor cell contamination in our samples, allowed to pass through the sort due to cell autofluorescence or low $\alpha$CD45 antibody specificity. 

To differentiate between these possibilities we examined library sizes of each cluster, reasoning that low-quality cells should have smaller libraries than high-quality counterparts. 
We found no association between library size and quality of immune cell correlation. 
We next re-ran the annotation after included several cell lines from epithelial and mesenchymal lineages as negative controls, reasoning that if the cell types were indeed stromal or cancer cells, they should appear closer in phenotype to these populations. 
This allowed us to identify several populations of fibroblasts, epithelial populations.
Post-hoc analyses of the epithelial populations by our surgeon allowed us to further differentiate the epithelial populations into malignant and non-malignant clusters, providing additional support for their separate clustering. % todo see if you can find a figure for this.  
Because these populations came primarily from only two patient samples, statistical power would be too low to derive meaningful conclusions about them, and they were excluded from downstream analyses. 

\section{Variation Between Individual Tumor Immune Microenvironments}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure1-D.png}
\caption{Pie charts showing cell type fractions in each patient tumor. 
}
\label{fig:1d}
\end{figure}

Having filtered out non-immune types and positively identified immune cells in each patient, we next examined the relative frequency of immune types across tumors.
In agreement with the mass cytometry analyses introduced in Chapter 1 \citep{Chevrier2017,Lavin2017} and prior clinical observations, we found a large degree of variation in the immune cell composition of each tumor (Figure~\ref{fig:1d}). % find a citation for the clinical knowledge
For example, the fraction of T cells varied between 21\%-96\% and the fraction of myeloid cells varied between 4-55\%. To determine the reliability of InDrop's sampling of these heterogeneous populations of immune cells, we compared the proportions of cell types as measured by flow cytometry and InDrop scRNA-seq. 
Although a comparison of the relative representation of major immune cell types identified by scRNA-seq to those measured by FACS revealed a significant bias towards Monocytic lineage cell subsets relative to expected input ratios, we observed high correlation between cell type frequencies across all patient samples ($r^2 > 0.8$, Figure~\ref{fig:s1c}). 
The observed bias, likely due to the larger cytoplasmic volume and higher RNA yield of Monocytic/myeloid cells vs. T cells, was systemic and did not adversely affect our analyses.
As a result, we concluded that we were able to identify the majority of immune cell types expected to be present in human tumors, including Monocytes, Macrophages, Dendritic cells, T cells, B cells, Mast cells, and Neutrophils (Figure~\ref{fig:1c},\ref{fig:s1c}) \citep{Jeffrey2006,Novershtern2011}. 
Thus, we were able to capture a comprehensive representation of the immune ecosystem from each individual tumor.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS1-C.png}
\caption{Regression of flow cytometry cell type percentages in each patient against RNA-seq cell type percentages for B cells (blue), Monocytic cells (orange), and T cells (green).
}
\label{fig:s1c}
\end{figure}

\section{Integration of Data Across Multiple Tumors}

To enable an unbiased systematic comparison across patients, we attempted to merge the data from all tumors to create a map of tumor-infiltrating immune cells. 
However, we observed that the normalization approaches applied to individual tumors were not adequate for data derived from multiple patients. 
Cells from the same patient, of different types, were often more similar than cells of the same lineage from another patient (Figure~\ref{fig:2a}). 
Figure~\ref{fig:2a} (left) shows scRNA-seq data from 9K immune cells from 4 breast cancer patients after normalization of cells to median library size, suggesting large differences between patients.
Moreover, the tSNE projection did not suggest a diversity of subpopulations beyond two main lymphoid and myeloid lineages. 
Since biological lineage should, in most cases, produce larger differences in transcript abundances than external signaling from the microenvironment, and we had already confirmed the presence of finer structure within individual patients, we believed that this phenomena was most likely due to technical effects acting across the different InDrop runs.
However, we also observed that activated immune cells contained higher numbers of mRNA molecules, a phenomena which has been previously reported \citep{Blackinton2016,Cheadle2005,Marrack2000,Singer2016}. 
Specifically, our analyses showed a gradient of activation of CD8 T cells in tumors as compared to normal- or blood-resident T-cells, where the most pronounced T-cell activation occurred in a TNBC tumor (BC3), which agrees with reports from clinical trials suggesting that TNBC tumors are the most immunogenic(Figure~\ref{fig:2b}) \citep{Dushyanthen2015,Garcia-Teijido2016}. 
Thus, our data displays technical and biological factors that both influence molecule abundance in individual samples. 

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-B.png}
\caption{Left: Boxplots showing expression of CD8 T cell activation signature (defined as the normalized mean expression of genes in the activation signature listed in Table S4) across immune cells from each patient. Right: Heatmap displaying z-scored mean expression of genes in activation signature. Top: Bar plot showing total expression of each gene indicated in the Heatmap across all patients. Expression of T cell activation signature shows variability across patients and increased expression in patients BC6 and BC3.} % need to figure out table citation
\label{fig:2b}
\end{figure}

% todo make this paragraph a bit cleaner. 
The tendency of samples to co-cluster highlights the complexity of analyzing scRNA-seq data from multiple patients.
There are various reasons, both technical and biologically stochastic, that result in co-clustering of samples. 
First, because the observed data is only one small sample set from the transcriptome of the cell (the full range of mRNAs that the cell expresses to support its phenotype) there is a high chance of missing low-expression genes. 
In addition, the depth of sampling strongly correlates with the number of features that are observed in each cell, and the sampling depth varies significantly across cells and across samples. % can we demonstrate this? 
The sparsity of scRNA-seq measurements mean that, in our data, the average gene is detected with only a single count. 
As a result, drop-out is very common, and drop-out is not recoverable by median library size normalization, as any number multiplied by zero remains zero. 
Thus, in small cells, detected genes are scaled up far more than they should be, producing spurious differential expression relative to better sampled cells of the same state. % todo image from biscuit presentation. 

Technical factors include differences in sequencing machine, enzyme activity, lysis efficiency or experimental protocol. 
These samples were also subject to operational variation during the clinical resection, transport, and handling. These factors all impact cell viability, which in turn affects the single cell RNA-seq library preparation, in particular molecular capture rate and sampling. 
Because molecular capture is a binary event, and the capture rates are very low, these technical variations often determine whether a given gene feature is observed in the data for a given cell.

These more technical artifacts, particularly in capture rate, are confounded with biological differences. 
This is particularly challenging in the case of immune cells, where activated cells have substantially heightened transcription rates, and therefore if sampling efficiency is constant, we expect to capture more molecules molecules \citep{Blackinton2016,Cheadle2005,Singer2016}. 
Therefore, the sampling rate is affected by biological as well as technical processes. 
Indeed, we see large differences in the number of activated T-cells across patients (Figure~\ref{fig:2b}), with more activated T-cells in the Triple Negative subtype as expected \citep{Dushyanthen2015}.
Hence, normalizing by library size will likely remove these biological variations.

Both the technical and biological effects tend to average out at the population level, but distance metrics do not share information across cells, as bulk approaches were able to..
As a result, distance metrics tend to be very sensitive to differences in sampling, which can lead to spurious differential expression or removal of biological stochasticity specific to each cell type, both of which induce improper clustering and characterization of latent cell types.
Therefore, cell type-specific normalization is especially crucial in experiments involving vast subtype diversity, such as immune cells ranging from large Macrophages to much smaller lymphocytes \citep{Lun2016,Vallejos2017}, wherein the sampling rate contains biological information. 

Unfortunately, cell's types are defined by the clusters they fall in, and thus cannot be not determined a priori. 
Thus, the transformation and clustering of scRNA-seq is a chicken-and-egg problem wherein it is illogical to start from either step.

\section{Biscuit Clustering and Normalization Corrects for Technical Effects Across Samples}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-A.png}
\caption{T-SNE projection of tumor-infiltrating immune cells from 4 breast cancer patients after library-size normalization (left panel) and Biscuit normalization and imputation (right panel). Cells are colored by tumor (patient). Less mixing of tumors indicates either batch effects or patient-specific cell states.
}
\label{fig:2a}
\end{figure}

To solve this problem we developed and applied the method ``Biscuit'' (short for Bayesian Inference for Single-cell ClUstering and ImpuTing) to simultaneously cluster cells and normalize according to their assigned clusters \citep{Prabhakaran2016} (Figure~\ref{fig:m2}).
To address the convolution of biological and technical effects, we instead model biological and technical variation in parallel, which we refer to as BISCUIT (Bayesian Inference for Single-cell ClUstering and ImpuTing).
This is accomplished through incorporating parameters denoting cell-specific technical variation into a Hierarchical Dirichlet Process Mixture Model (HDPMM) \citep{Goeruer2010} (Figure~\ref{fig:m2}-A). 
This allows for inference of cell clusters based on similarity in gene expression as well as in co-expression patterns, while identifying and accounting for technical variation per cell (Figure~\ref{fig:m2}B,C). 
Two key ideas that power Biscuit are the use of gene co-expression as a more robust means to identify cell types, and the normalization of each cell type separately to better account for cell type-specific effects on technical variation. 
The main idea behind the use of co-expression is that cell types not only share similar mean expression, but also share similar co-expression patterns (covariance) between genes. 
While mean expression can be more sensitive to capture efficiency, covariation is more robust to such effects. 
This similarity in co-variation can be used to improve normalization and in turn improve the clustering, through the learning of cluster specific parameters.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureM2.png}
\caption{(A) Stochastic data generative process for Biscuit illustrated with a toy example. Top panel: Left: shows 3 multivariate Gaussian densities with no technical variation. Middle: An ideal cell ($y_j$) is simulated as a random draw from any of these 3 Gaussians. Right: The covariance matrix across 10 such randomly-drawn cells showing 3 block covariances across the diagonal corresponding to three clusters.  
  Bottom panel: Left: shows 3 multivariate Gaussian densities with means and covariances scaled using ($j, j$) to handle cell-specific variations. 
  Middle: A cell ($l_j$) is simulated as a random draw from any of these 3 scaled Gaussians. 
  Right: The covariance matrix across 10 such randomly-drawn cells showing loss of signal in the 3 block diagonal covariances. 
  We assume the model for $l_j$ captures real single-cell measurements and the goal is to normalize data by converting it to follow the model for $y_j$. 
(B). Finite state automata for Biscuit. 
The shaded circle denotes $l_j$, which is observed gene expression for cell $j$, white circles show latent variables of interest, rectangles depict the number of replications at the bottom right corner, diamonds are hyper-parameters, and double diamonds are hyper-priors obtained empirically. Inference equations are obtained by inverting the date generative process. 
(C) Left panel: Input count matrix to Biscuit. Middle panel: Inference algorithm with Gibbs iterations are depicted where cell-specific ($j, j$) and cluster-specific ($k, k$) parameters are iteratively inferred leading to cell assignments to clusters. Right panel: Output from Biscuit, which is the normalized and imputed count matrix.
}
\label{fig:m2}
\end{figure}

By jointly performing normalization and clustering, we retain biological heterogeneity and avoid biases that result from independent clustering and normalization, and instead are able to match cells to clusters of the same cell type from different patients which may have very different sampling rates. 
Figure~\ref{fig:2a}(right) shows the same data from 4 tumors after normalization with Biscuit. 
Note that Biscuit does not use any information on sample IDs in the normalization, and normalization is only driven by cluster assignments. 
The Biscuit-normalized data shows that the differences in library-size normalized data were largely artifacts of normalization and batch effects. 
We therefore applied Biscuit to data from all 8 tumors to infer the full diversity of immune cell types in the breast tumors, which identified 67 clusters indicating significant diversity in both lymphoid and myeloid cell types (Figure~\ref{fig:2c}). % fix figure reference

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-C.png}
\caption{T-SNE map of breast tumor-infiltrating immune cells from all 8 patients after Biscuit normalization and imputation showing rich structure and diverse cell types. Cells colored by Biscuit clusters and labeled with inferred cell types.
}
\label{fig:2c}
\end{figure}

This transformation also imputes dropouts in each cell by sampling dropped-out genes from the posterior distributions for the cluster that a cell is assigned to. 
The use of covariance parameter in the model ensures that intra-cluster heterogeneity is preserved after imputing. 
We show a systematic evaluation of the algorithm performance (on synthetic and real single cell data), its robustness, as well as the ability of this method to impute dropouts in \citep{Prabhakaran2016}.

% this may need to be moved up
To formalize and quantify Biscuit's ability to correct batch effects across data from all eight tumors (Figure~\ref{fig:2c}) and match immune subtypes across the tumors, we devised an entropy-based metric that quantifies the ``mixing'' of the normalized data across samples. 
The entropy-based metric is computed as follows: We constructed a k-NN graph (k=30) on the normalized data using Euclidean distance and computed the distribution of patients (tumors) \(m = 1,\ldots,8\) in the neighborhood of each cell \(j\), denoted as \({q_{j}}^{m}\). 
Then we computed Shannon entropy \(H_{j} = \  - {q_{j}}^{m}\text{\ log}{\text{\ q}_{j}}^{m}\) as a measure of mixing between patients, resulting in one entropy value \(H_{j}\)per cell \(j\). 
High entropy indicates that the most similar cells come from a well mixed set of additional tumors, whereas low entropy indicates that the most similar cells largely come from the same tumor. 
Prior to Biscuit, most cells in the data had low entropy values, with 40\% of the cells residing in a neighborhood of cells purely from the same tumor.
We compare the distribution of entropies across all cells from all 8 tumors, before and after Biscuit, which reveals that the median of entropy shifts significantly towards higher mixing of samples after processing with Biscuit (Mann-Whitney U-test: U=1.7721e+09, p=0; Figure~\ref{fig:2d}). 
Thus, we conclude that Biscuit substantially corrected batch effects in this data.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-D.png}
\caption{Histogram depicting entropy of the patient distribution as a measure of sample mixing. Entropy is computed per cell, based on the distribution of patients in (30-NN) local cell neighborhoods after library-size normalization (left panel) as compared to Biscuit (right panel).  
}
\label{fig:2d}
\end{figure}

To generate a global atlas of immune cell types, we combined samples from all patients and tissues by applying Biscuit to the full set of $n=62024$ cells and $d=14875$ genes, resulting in a global atlas of $K=95$ clusters (Table S2) in which $n=57143$ cells had statistically significant cluster assignments.  % may need to address what this means. 
The remainder of cells had low library size and were hence removed from further analysis. % reference table 
A subset of these clusters were identified as probable cancer or stromal populations through correlation with bulk gene expression datasets and marker gene expression. 
While these non-immune clusters may be of significant interest in their own right, they were beyond the scope of this paper and were therefore excluded from downstream analysis, leaving 47,016 cells in 83 clusters (Table S2). % reference table

While biscuit improved mixing across patients overall, We observed that individual clusters displayed differing amounts of mixing between samples (Figure~\ref{fig:s2g}).
To further quantify the exact degree of mixing (between patients) in each cluster, we defined an entropy-based metric.
We used bootstrapping to correct for cluster size (which ranged from over 8900 cells to just over 30 cells), such that we uniformly sampled 100 cells with replacement from each cluster, computed the distribution of patients across these cells, and then computed the Shannon entropy for this distribution.
We repeated this procedure 100 times for each cluster, to achieve a range of entropy values per cluster.
Figure~\ref{fig:s2d} shows box plots for entropy values for each cluster, with the order of clusters based on their mean entropy.
Clusters with entropy of 0 denote entirely patient-specific clusters.
Figure~\ref{fig:s2d} shows that there is a continuous range of entropies, and thus a full range of sample specificity versus mixing, across clusters.
These results suggest that our experiment succeeded in observing both general immune cell states, and also tumor-specific states that may result from specific microenvironments. 

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-E.png}
\caption{T-SNE projection of complete atlas of immune cells, post-Biscuit normalization, from all patients and all tissues including tumor, blood, lymph, and contra-lateral normal tissue,labeled by inferred cell type (left panel) and normalized expression of 8 immune cell markers (right panel). Figure S2 presents further details on inferred clusters with complete annotations in Table S2.
} % reference the table
\label{fig:2e}
\end{figure}

\section{Cluster Robustness}

To evaluate cluster robustness, we performed 10-fold cross-validation, independently clustering and normalizing on random subsets of data.
For each of 10 subsets, we ran Biscuit to obtain a set of clusters.
To compare the results across the 10 subsets, we computed the confusion matrix, which indicates the probability of each pair of cells \(j,j'\) being assigned to the same cluster: \(P(z_{j} = z_{j'})\), where $z_j$ is the $j^{th}$ cell.
Figure~\ref{fig:s2c} illustrates box plots for the probabilities of co-clustering (across 10 subsets) for every pair of cells that are assigned to the same cluster in the analysis of the full dataset.
The average co-clustering probability in each cluster ranges between 92\%-100\%, showing remarkable robustness of clusters.

\section{Distances Between Clusters}

The distances between BISCUIT clusters can be directly computed from the posterior probability distributions of each cluster. 
While Euclidean distances are defined for vectorial objects and operate under a Cartesian coordinate system, Euclidean distance with non-vectorial objects such as probability distributions requires embedding them in Euclidean space.
Such embeddings are non-unique and lead to loss of information.
It is therefore advisable to use the non-vectorial objects as is and to work with the objects' pairwise similarities or distances instead.
One such distance metric, which is effective at comparing pairwise probability distributions, is the Bhattacharyya distance (BD) \citep{Bhattacharyya1990}.

We defined distances between each pair of clusters \(k,k'\) with distributions \({p_{k}}^{}\) and \({p_{k'}}^{}\)as \(BD\  = - log(BC({p_{k}}^{},{p_{k'}}^{\ }))\) where \(\text{BC}\) is the Bhattacharyya coefficient measuring similarity (overlap) of the distributions.
We use the BD to compute distances between pairs of inferred clusters' moments to create the Bhattacharyya kernel.
The Bhattacharyya kernel has closed forms for any exponential distribution including the (multivariate) Gaussian distribution \citep{Jebara2004}, which is Biscuit's underlying data-generation distribution.
For the case of multivariate normal distributions: \({p_{k} \sim N(\overrightarrow{\mu_{k}},\Sigma_{k})}^{}\)and \({p_{k'} \sim N(\overrightarrow{\mu_{k'}},\Sigma_{k'})}^{}\):

\(D_{B}\  = \frac{1}{8}(\overrightarrow{\mu_{k}} - \overrightarrow{\mu_{k'}})^{\text{T\ }}\Sigma^{- 1}(\overrightarrow{\mu_{k}} - \overrightarrow{\mu_{k'}})\  + \ \frac{1}{2}log(\frac{\text{det\ }\Sigma}{})\ \) where \(\Sigma = (\Sigma_{k} + \Sigma_{k'})/2\).

Figure~\ref{fig:s2b}  shows the Heatmap of pairwise distances between all pairs of
clusters.

A geometric interpretation of BD is that, via its cosine formulation, the distance subsumes a full hypersphere and the centre of the hypersphere is the centroid (mean) of the cluster, whereas the Euclidean distance only covers a quarter of the hypersphere with the center at the origin. % todo why is this important? ask elham

\section{Contribution of Covariance in Defining Clusters}

We used the above Bhattacharyya distance (BD) metric to study the contribution of Biscuit's covariance parameters to characterizing clusters of different cell types.
First, we computed the BD between pairs of clusters of the same type (T, Monocytic, NK, B) and compared these to distances between pairs of clusters of different cell types (e.g.\ a T cell cluster and a Monocytic cluster).
Figure S2E shows violin plots for distances between pairs of clusters with dots (overlaid on violins) representing cluster pairs; violins are sorted based on median distance.
As reference, we also split each cluster into two halves and computed the empirical BD between two splits (shown at the left end in Figure~\ref{fig:s2e}).
We observe that, overall, pairs of clusters of different types are more distant than pairs of clusters from the same type, as expected.

We then computed these same pairwise distances while removing the contribution of mean parameters for each cluster, via setting \(\overrightarrow{\mu_{k}} - \overrightarrow{\mu_{k'}} = 0\) and computing the distance only based on covariance parameters of the pair of clusters \(\Sigma_{k},\Sigma_{k'}\) (Figure~\ref{fig:s2e}, right).
We observed that pairs of T cell clusters or Monocytic clusters still show prominent distances, and therefore covariance parameters have a crucial role in defining these clusters.

% RECONCILE THIS WITH THE ABOVE -- WHAT FOLLOWS IS A BIT OF THE START OF A BIOLOGICAL ANALYSIS. COULD
% MOVE TO NEXT CHAPTER IF NEEDED. 

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-F.png}
\caption{Pearson correlations of cluster expression centroids to bulk RNA-seq data from purified immune populations (from Jeffrey 2006 and Novershtern 2011). Scale bar displays r-values.
(G) Histogram of frequency of patients contributing to each cluster showing that 19 clusters (out of 95) are present in all 8 patients and 10 clusters are patient-specific.
}
\label{fig:2f}
\end{figure}

In the case of Biscuit clusters (Figure~\ref{fig:2f}), mean parameters for each cluster were correlated with bulk profiles.
Each of the bulk profiles was marked as having derived from one of several major cell types: b-cells; T-cells (naive, central memory, cytotoxic, T-regulatory); Monocytic cells (monocytes, dendritic cells, macrophages); Mast cells; Neutrophils; or NK-cells.
The highest scoring bulk profile for each centroid was used to categorize each cluster by its type, and types were split for downstream analysis.

% THIS HAS ALREADY BEEN INTRODUCED ABOVE, MAKE THIS SHORTER. 
Cells were also typed by examining expression of known marker genes.
In this analysis, cells were scored as detecting a marker gene if the cell contained a non-zero molecule count for that gene.
Each cell was corrected for its detection rate (the fraction of total genes detected in that cell) and the marker detection rate was then averaged across cells of a cluster.
Markers used to assign classical types cells included NCAM1, NCR1, NKG2 (NK-cells), GNLY, PFN1, GZMA, GZMB, GMZM, GZMH (cytotoxic T-cell, NK), FOXP3, CTLA4, TIGIT, TNFRSF4, LAG3, PDCD1 (exhausted T-cell, T-regulatory cell), CD8, CD3, CD4 (T-cells), IL7R (naive T-cells), CD19 (b-cells), ENPP3, KIT (Mast cells), IL3RA, LILRA4 (Plasmacytoid Dendritic cells), HLA-DR, FCGR3A, CD68, ANPEP, ITGAX, CD14, ITGAM, CD33 (Monocytic lineage).
For all retained clusters, the two typing methods agreed (Figure~\ref{fig:2h}).

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-H.png}
\caption{Expression of canonical and cell type markers across clusters, z-score normalized across clusters. T-exhausted denotes the mean expression of terminal differentiation signature listed in Table S4.
} % note that the table will need a reference in the text!
\label{fig:2h}
\end{figure}

\section{Gene Signature Summarization}

To interpret the observed cell states we made extensive use of gene signature enrichment.  
However, in addition to testing for heightened expression of genes in the signature across cells in the cluster, we also examined signatures in terms of their variation.
Specifically, we examine the marginal distribution of cell loadings across the signature and the relative contribution of each gene.

Therefore, when examining signature expression across patients we began by constructing a bar plot of the counts for each gene in the signature, corrected for cellular observation rate (the total number of genes observed with molecule count \textgreater{} 1).
This displays the contribution of each gene to the signature (top panel in Figure~\ref{fig:1e},\ref{fig:s1d}).
The normalized values for each signature, per cell, are then summarized as a box plot to display the variation of cells in each patient (left panels).
Finally, the cluster median of each gene is taken per patient, and the cluster medians are z-scored across patients.
The z-scored values are plotted as a Heatmap (center-right panel in Figure~\ref{fig:1e},\ref{fig:s1d}), facilitating a comparison of signatures across patients\footnote{The full lists of compiled gene signatures used throughout this dissertation can be found in Table S4 of the bioArxiv publication and may serve as a valuable resource for additional investigations. % # todo need a link from this.  
To create these lists we broadly surveyed the extant literature and manually curated consensus lists of genes to be included.
The relevant literature surveyed to form these signatures includes the following:

For the M1 and M2 macrophage polarization signatures we merged gene lists from \citep{Sica2012}; \citep{Biswas2010}; \citep{Bronte2016} \citep{Ugel2015} \citep{Gabrilovich2017}.
For other myeloid-specific signatures we used \citep{Villani2017} (pDCs, AXL/SIGLEC6 DCs, CD141/CLEC9A DCs, CD11C\_A DCs, CD1C-/CD141- DCs, CD1C\_B DCs, New Monocytes 1, New Monocytes 2, CD14+CD16- Classical Monocytes, and CD14+CD16+ Non-Classical Monocytes); and \citep{Gesta2007}, \citep{Perera2006}, \citep{Farmer2006,Lefterova2009} (Lipid Mediators).

For T-cell-specific signatures we used \citep{Wherry2015}, \citep{Wherry2011}, \citep{Schietinger2012} (Exhaustion and Anergy); \citep{Glimcher2004} (Cytolytic Effector Pathway); and \citep{Smith-Garvin2009}, \citep{Chtanova2005}, and \citep{AdamBest2013} (T-cell Activation).

For gene signatures used across cell types we used \citep{Mantovani2008}, and \citep{Grivennikov2010} (Pro and Anti-Inflammatory); \citep{Platanias2005} (Type I and II Interferon Responses); \citep{Ho2015} (glucose deprivation); \citep{Benita2009,Makino2003} (Hypoxia/HIF Regulated); \citep{Moreno-Sanchez2009}, \citep{Caton2010,Funes2007,Mues2009}, \citep{Beale2007} (Glycolysis, Gluconeogenesis, TCA Cycle, Pentose Phosphate Pathway, and Glycogen Metabolism), and \citep{Whitfield2002} (G1/S).}.


% todo needs an impact sentence about why we care about these additional things that we're doing. 

The genome-wide view allowed us to assess system-level differences between immune cell consortia in individual patients in, for example, metabolic signatures, including hypoxia (Figure~\ref{fig:1e}).
It is interesting to note that while all tumors expressed a similar average degree of a hypoxia signature, patients differed considerably in expression at the level of individual genes included in the signature.
Similar variation was observed in fatty acid metabolism, glycolysis, and phosphorylation (Figure~\ref{fig:s1d}).

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure1-E.png}
\caption{Left: Boxplots showing expression of Hypoxia signature (defined as the mean normalized expression of genes in the hypoxia signature listed in Table S4) across immune cells from each patient.
  Right: Heatmap displaying z-scored mean expression of genes in hypoxia signature.
  Top: Barplot showing total expression of each gene indicated in the Heatmap, across all patients.
  See Figure~\ref{fig:s1d} for additional signatures.
}
\label{fig:1e}
\end{figure}

\section{Breast Tumor Immune Cell Atlas Reveals Substantial Diversity of Cell States}

\begin{figure}
\centering
\includegraphics[width=0.75\textwidth]{Figure1-A.png}
\caption{Flow chart displaying experimental design and analysis strategy.}
\label{fig:1a}
\end{figure} % make sure suppl figure reference matches.

% THIS PART OF THE ANALYSIS WAS ALREADY SUMMARIZED ABOVE. WORK IT IN. 
Together, these analyses produce a tumor immune atlas that can be interrogated to understand tumor, tissue, and patient dependent differences in immune phenotypes.
A complete map of the experimental procedure described in this chapter and the one preceding it, from sample extraction to the end of BISCUIT clustering, is displayed in Figure~\ref{fig:1a}.
After applying Biscuit to the data from all tumors (Figure~\ref{fig:2b}), we found 67 clusters covering various T cell, macrophage, monocyte, B cell, and NK cell clusters.~We first asked whether individual cells tended to be most similar to cells from their own samples or if the resulting cell profiles were well mixed using an entropy measure (STAR Methods).
For each cell, this measure considers the neighborhood of its most similar cells and evaluates the entropy of the sample distribution in each such neighborhood.
Low entropy indicates that most neighbors come from the same sample, whereas high entropy indicates that the neighbors (most similar cells) are well distributed across the different samples.
Indeed, while cells were most similar within individual samples before normalization, this was corrected after Biscuit normalization with significantly improved mixing of cells across patients when compared against standard normalization methods (Figure~\ref{fig:2a},\ref{fig:2d}).
(U=1.7721e+09, p=0).
Using this approach, we successfully retained information on immune cell activation while stabilizing differences in library size, and uncovered a rich and robust structure in imputed data, suggesting diversity in immune cell subtypes (Figure~\ref{fig:2a},\ref{fig:2c}).

To construct a global atlas of immune cells, enabling characterization of the impact of environment on immune cell states, we merged data from 47,016 cells across all tissues and patients revealing a diverse set of 83 clusters, each identifying a cell type or state (Figure~\ref{fig:2e},\ref{fig:2f}).
This unexpectedly large number of clusters prompted us to test their robustness using cross-validation on subsets of the data (STAR Methods), finding assignments of cells to clusters were robust for most clusters (Figure~\ref{fig:s2c}).
Most clusters were shared across multiple patients, indicating similar immune states across patients, with only 10 being patient-specific (Figure~\ref{fig:s2g}).
We used entropy as a more stringent metric for patient mixing within clusters and found that the clusters span a range of different mixing levels (Figure~\ref{fig:s2g}).

We assigned each cluster to its associated cell type by comparing cluster mean expression to bulk RNA-seq as described above (Figure~\ref{fig:2e},\ref{fig:2f}) and found 38 T cell clusters, 27 myeloid lineage clusters, 9 B cell clusters, and 9 NK cell clusters (Table S2). % fix table reference
By examining the expression of canonical markers in immune cell clusters, we were able to confirm and build upon predictions made by the preceding analysis (Figure~\ref{fig:2h}).
Of the T cell clusters, we identified 15 CD8+ T cell clusters and 21 CD4+ T cell clusters, which were together split into 9 naive, 7 central memory, 15 effector memory, and 5 Treg clusters.
We were additionally able to divide the myeloid lineage clusters into 3 macrophage, 3 mast cell, 4 neutrophil, 3 dendritic cell, 1 plasmacytoid dendritic cell, and 13 Monocytic clusters.
Finally, we identified 9 B cell clusters, 3 CD56\textsuperscript{-\/-} NK cell clusters, and 6 CD56\textsuperscript{++} NK cell clusters, of which 2 of which are likely NKT cells.
These clusters can be distinguished by their differential expression patterns (Figure~\ref{fig:2j}).

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-J.png}
  \caption{Differentially expressed genes in b-cells (top) and NK-cells (bottom) standardized by z-scores within cell type.
  As an example, the expression of CD19 is standardized across all B cell clusters to highlight clusters with higher or lower expression of the marker compared to the average B cell cluster, but is highly expressed in nearly all B cell clusters (refer to Table S3 for all DEGs in these and other clusters).
  % proper reference to table
}
\label{fig:2j}
\end{figure}

Since our characterization identified multiple clusters with the same cell type ``label'' based on surface markers and prior characterization of the corresponding peripheral blood cell phenotypes, e.g.  15 effector memory T-cell clusters (Figure~\ref{fig:2f}), we wanted to confirm that all these clusters were indeed distinct.
The distributions defined by the Biscuit parameters identified differentially expressed genes between clusters, including canonical immune genes, and defined multiple subpopulations within each major cell type (Table S3). % todo fix table citation
Moreover, we observed a prominent effect of covariance in defining the T cell clusters by comparing similarity of pairs of clusters with and without the effect of mean expression (Figure~\ref{fig:s2e}); large differences between most clusters remained even after mean gene expression was equalized.
Thus, our approach robustly identified cell states that were distinct from one another and shared across multiple tumor microenvironments.
As T-cell and myeloid cells represent the most abundant and diverse, and arguably most biologically significant, immune cell subsets in the tumor microenvironment, we focused our subsequent in-depth analyses on these two major cell types.
This investigation is detailed in the next, final chapter of the dissertation. 

