
\chapter*[Supplementary Figures]{Supplementary Figures}
\addcontentsline{toc}{chapter}{Supplementary Figures}
% \refstepcounter{chapter}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{platform.png}
\caption{Code for the 10x platform class. Each platform class defines the length of the primer sequence, a merge function to combine the barcodes and the genomic fastq, and optionally, custom barcode correction and multi-alignment resolution methods. 10x uses the default multi-alignment method, and therefore that does not require implementation in this class.}

\label{fig:10x_platform}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS1-B.png}
\caption{t-SNE projection of complete immune systems from six breast cancer tumors.\ scRNA-seq data for each tumor is processed with pipeline in Figure S1B and library size-normalized; each dot represents a single- cell colored by PhenoGraph clustering, and clusters are labeled by inferred cell types. Two additional tumors are presented in~\ref{fig:1c}.
}
\label{fig:s1b}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=0.75\textwidth]{FigureS1-D.png}
\caption{Expression of metabolic signatures: fatty acid metabolism (top), phosphorylation (middle), and glycolysis (bottom), summarized as boxplots (left) showing expression of each respective signature (defined as the mean normalized expression of genes)  across immune cells from each patient; and heatmap (right) displaying z-scored mean expression of genes in each signature; (top) barplot showing total expression of each gene indicated in the heatmap across all patients. See Figure~\ref{fig:1e} for one additional signature
}
\label{fig:s1d}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-A.png}
\caption{Posterior probability of assignment of cells to clusters in the Biscuit model in the full immune cell atlas of combined tissues and patients presented in Figure~\ref{fig:2e}; note broad distributions in assignment of naive T cells (bottom) as compared to other cell types.
}
\label{fig:s2a}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-F.png}
\caption{Distribution of Biscuit alpha parameters per cell vs log of library size, with cells colored by clusters; Biscuit alpha parameters correct for differences in library size across and within clusters.
}
\label{fig:s2f}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-G.png}
\caption{Distribution of inferred cell-specific parameters alpha and beta in Biscuit across cells from each patient. These differences were corrected in normalizing with alpha and beta parameters.
}
\label{fig:s2g}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-C.png}
\caption{Robustness analysis of clusters performed with 10-fold cross-validation; boxplots summarize the probability of a pair of cells being assigned to the same final cluster across all 10 subsets.
}
\label{fig:s2c}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{Figure2-G.png}
\caption{Histogram of frequency of patients contributing to each cluster showing that 19 clusters (out of 95) are present in all 8 patients and 10 clusters are patient-specific.
}
\label{fig:2g}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=0.75\textwidth]{FigureS2-D.png}
\caption{Boxplots showing entropy of distribution of patients in each cluster, computed with bootstrapping to correct for cluster size. Note that cluster labels are given by size (cluster 1 has the most number of cells and cluster 95 has the fewest) and ordering clusters by mean entropy in this plot indicates that entropy does not correlate with size.
}
\label{fig:s2d}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-B.png}
\caption{Bhattacharyya pairwise distances between clusters of Figure~\ref{fig:2h} (blue: small distance to yellow: large distance).
}
\label{fig:s2b}
\end{figure}


\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS2-E.png}
\caption{Left: Violin plot of pairwise Bhattacharyya distances between distribution of expression of each gene between all pairs of clusters in the same or different cell types considering mean and covariance of expression, averaged across all genes.
Right: same as left, but after removing the effect of cluster mean in computing similarity, thus considering only covariance.
}
\label{fig:s2e}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS3.png}
\caption{Hallmark GSEA enrichment results on genes with highest difference in variance in tumor vs normal tissue in (A) NK and (B) monocytic cells. See Figure~\ref{fig:3d}  for enrichment in T cells; complete lists of enrichments are presented in Table S5.
}
\label{fig:s3}
\end{figure}


\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS4-A.png}
\caption{Hartigan’s dip test on density of cells projected on diffusion components, showing statistically significant continuity (lack of “dips”) in cells along T cell activation component (component 3, third panel from left), whereas other components exhibit more defined states (multimodality).
}
\label{fig:s4a}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS4-B.png}
\caption{Violin plot of cells projected on terminal differentiation diffusion component: component organized by tissue type (left panels) and cluster (center panel). Also, heatmap showing expression of immune-related genes with the largest positive correlations with component, averaged per cluster and z-score standardized across clusters; columns (clusters) are ordered by mean projection along the component.
}
\label{fig:s4b}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS4-D.png}
\caption{Violin plot of cells projected on hypoxia diffusion component: component organized by tissue type (left panels) and cluster (center panel). Also, heatmap showing expression of immune-related genes with the largest positive correlations with component, averaged per cluster and z-score standardized across clusters; columns (clusters) are ordered by mean projection along the component.
}
\label{fig:s4d}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS5-A.png}
\caption{Displaying null distributions and observed covariances between CTLA-4 and GITR in raw, un- normalized data using hypothesis testing, subsampling, and permutation (see STAR methods); shows that the differences in covariance shown in biscuit-normalized data are also present in un-normalized and un-imputed data, and hence are not an artifact of computation.}
\label{fig:s5a}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS5-B.png}
\caption{Bivariate plots of expression levels of GITR and CTLA-4 in Treg clusters based on inferred mean and covariance parameters from Biscuit. Dark blue color indicates the highest density of cells and light yellow the lowest density of cells.
}

\label{fig:s5b}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS5-C.png}
\caption{Network graphs showing covariance between checkpoint receptors in activated T cell clusters. Edge width denotes absolute magnitude (strength) of covariance and color denotes sign of covariance (red positive and blue negative). Note diversity across clusters. 
}

\label{fig:s5c}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS6-A.png}
\caption{Hartigan’s dip test on density of cells projected on diffusion components indicating no diffusion components across myeloid cells show statistically significant continuity, implying myeloid cells reside in defined (multimodal) states along major components explaining variation.}
\label{fig:s6a}
\end{figure}

\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS6-B.png}
\caption{Violin plot showing the density of cells projected along pDC component and organized by tissue type and cluster.
}
\label{fig:s6b}
\end{figure}


\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS6-C.png}
\caption{Violin plot showing the density of cells projected along monocyte activation component and organized by tissue type and cluster.
}
\label{fig:s6c}
\end{figure}


\begin{figure}
\centering
\includegraphics[width=\textwidth]{FigureS7-A.png}
\caption{Displaying null distributions and observed covariances between MACRO and CD276 in raw, unnormalized data using hypothesis testing, subsampling, and permutation (see STAR methods), showing that the differences in covariance in normalized data as shown in  Figure 7B are also present in un-normalized and un-imputed data, and hence is not an artifact of computation.
Bivariate plots of expression levels of MARCO and CD276 in Treg clusters based on inferred mean and covariance parameters from Biscuit. Dark blue color indicates the highest density of cells and light yellow the lowest density of cells.
}
\label{fig:s7a}
\end{figure}

