%This is the conclusion of the dissertation

\chapter*{Discussion and Perspectives} 
\addcontentsline{toc}{chapter}{Discussion and Perspectives}

\section{Implications of the Breast Tumor Immune Atlas}

Despite major clinical advances in cancer immunotherapy, our ability to understand its mechanisms of action or predict its efficacy is confounded by the complex, heterogeneous, and poorly understood composition of immune cells within tumors. 
Since cancer is generally a disease that affects older, post-reproductive individuals, with the exception of inherited genetic predisposition, it is unlikely that specialized mechanisms of the adaptive or innate immunity evolved to facilitate tumor immune surveillance. 
It seems reasonable to suggest that immune mechanisms affecting tumor progression must also operate in non-cancerous tissues to maintain organismal integrity and tissue function in the face of infection, stress, inflammation, and injury. 
A corollary to this notion is that features of immune cells in tumors must, by and large, resemble features of cells in non-cancerous tissues. 
A recent population-level RNA-seq analysis of T Reg cells and effector CD4 T cells in breast cancer and normal breast tissue identified a high level of phenotypic similarity between tissue and tumor-resident T cells, thus providing experimental support for this idea (Plitas et al., 2016). A similar RNA-seq study focusing primarily on T Reg cell analysis in colorectal and lung cancer suggested that cancer-resident T Reg cells differ considerably from those found in the normal tissue (De Simone et al., 2016).  % fix citations
Despite seeming differences in conclusions, distinguishing features of intra-tumoral T Reg cells as compared to normal tissue-resident ones detected in these two reports were associated with their heightened activation and thus can not distinguish between differences in immune states themselves or differences in immune state proportions. 
Thus, the averaging of gene expression features in the bulk cell population analyses and the lack of the assessment of a broad spectrum of immune cell subsets do not allow for a definitive investigation of specific effects of the tumor environment on immune cells. 

To address this question, we characterized the available single-cell approaches and selected and adapted InDrop, the best-suited single-cell method to assaying rare populations of immune cells. 
With InDrop, we undertook an unbiased comparative single-cell RNA-seq analysis of all tumor versus normal tissue-resident immune cell subsets and constructed an comprehensive immune atlas in breast carcinomas, combining immune cells isolated from normal and cancerous breast tissue, as well as peripheral blood and the lymph node. 
Our analysis was empowered by a suite of novel computational tools for single-cell RNA-seq data, including a data processing pipeline more sensitive in its ability to detect immune molecules, a powerful clustering and normalization algorithm, and new metrics for volume of the phenotypic space.  
These secondary analysis methods allowed us to overcome significant technical artifacts, correct for amplification biases, eliminate spurious molecules generated by library construction, recover molecules that would otherwise be hidden by multiple alignment, and select and eliminate problematic cells that were undergoing apoptosis, expressing transcripts consistent with technical stressors, or whose profiles represented undesirable cell types that had escaped flow sorting. 
Our novel tertiary methods facilitated clustering and normalization in the face of the strong batch effects typical of clinical samples that would have otherwise dominated the signal and obscured the identification of shared cell states across tumors. 
Additionally, the BISCUIT model facilitated an in-depth characterization of variance within and between identified cell states, a goal that had been at best partially realized in single-cell analysis. 

The constructed atlas revealed vast diversity in the repertoire of immune cells representative of both the adaptive and innate immune systems. 
Our examination of hematopoietic nucleated cells from treatment-na√Øve human breast cancer and normal breast tissue across different patients revealed that the biggest change to the immune cells was linked to the tissue environment, resulting in cell states that are substantially different than those present in the blood and lymph node. 
Interestingly, immune cell subpopulations in normal tissue were observed to be a subset of those found in tumor tissue, an observation that could not have been found with bulk gene expression measurements. 
Furthermore, the diversity of cell states significantly expanded between normal tissue and tumor, as quantified with our metric of ``phenotypic volume'' occupied by immune cell states. 
We observed tremendous expansion of the immune phenotypic space occupied by all major cell types in breast tumors as compared to normal breast tissue. 
It seems reasonable to speculate that the majority, if not all immune cell states found in cancer can be found in corresponding non-cancerous tissues in response to different stresses such as infection, wound healing, or inflammation.

The observation of an expanding continuous T cell ``phenotypic space'' in the tumor argues against the view of activated T cells rapidly traversing through sparse transitional cell states towards a few predominant, discrete, and stable states, including T Reg, effector, memory, and exhausted T cells. 
Three major components contributed to this phenotypic expansion in tumor tissue that helped explain the heterogeneity of T cells contributing to this phenotypic expansion in tumor tissue, we identified three major components that help explain the heterogeneity of T cells, including T cell activation, terminal differentiation, and hypoxic response. 
The strongest of these components is a predominant trajectory of progressive T cell activation and differentiation across 38 T cell clusters, including T Reg and terminally differentiated T cell clusters, found at the extreme activation terminus. 
One obvious explanation for the ``continuity'' of intra-tumoral T cell activation is the presence of increasingly diverse environments defined by a multitude of gradients including growth, pro-inflammatory, and tissue repair factors, as well as oxygen, nutrient, and metabolite gradients which exist to a lesser extent in healthy breast tissue (Buck et al., 2017).  % ifx citations
Indeed, we found groups of genes within the corresponding signaling pathways, most prominently immune activation (IFN/IL6/JAK/STAT) and hypoxia, to be differentially expressed across T cell clusters.

A non-mutually exclusive possibility is that the wide range of TCR signal strengths afforded by a diverse repertoire of T cell receptors (TCR) accounts for the continuous spectrum of T cell activation, obscuring the transitional states. 
The latter may also be accounted for by asynchrony in polyclonal T cell activation or heterogeneity in the types of antigen-presenting cells, their activation status, and their anatomical distribution. 
Unlike polyclonal T cell populations, activation of a monoclonal T cell population with a ``fixed'' specificity for tumor ``self'' or neo-antigen may yield sparse discontinuous ``phenotypic'' spaces reflecting discrete functional T cell states. 
In support of the latter possibility, recent bulk gene expression and chromatin accessibility analyses showed that cognate tumor neo-antigen recognition by TCR transgenic T cells results in an orderly progression of activated T cells through a reversible dysfunctional intermediate state towards an irreversible dysfunctional terminal state (Philip et al., 2017). % fix citations 
Additionally, diverse TCR specificities are known tocan contribute to spatial distribution of T cells and, therefore, facilitate their exposure to the distinct environments (``mini-niches'') discussed above. 

While T cells of various cell types exhibit continuous levels of activation, our inferred subsets further show variable levels of responses to environmental stimuli, and the combinations of these environmental exposures jointly define the identity of discrete CD4+/CD8+ T cell subsets. 
We also identified 5 T Reg subsets that showed similar responses to environmental pressures and shared differentially expressed genes, but exhibited drastic differences in gene covariance patterns. 
Particularly noteworthy was co-expression of checkpoint receptor genes in some T Reg subpopulations as compared to mutually exclusive expression of the same genes in other T Reg clusters. 
In this regard, co-variant expression of CTLA-4, TIGIT and co-stimulatory receptor GITR and other co-receptors in multiple T Reg cell clusters suggests that these T Reg cell populations may occupy different functional niches; CTLA-4 and TIGIT co-expressing cells have been demonstrated to selectively inhibit pro-inflammatory Th1 and Th17, but not Th2 responses promoting tissue remodeling (Joller et al., 2014). % fix citations
The observed co-expression of functional cell surface and signaling molecules by intra-tumoral T Reg cells may enable targeted modulation of T Reg cell activity in the tumor microenvironment using combinatorial therapeutic approaches (Mantovani and Locati, 2013).  % fix citations
This finding has implications in the way we describe and interrogate the tumor immune response. 
It is also noteworthy that the discrete cell states that are commonly utilized to describe immune responses are largely defined from highly polarizing conditions such as infection and tissue injury.

Our analyses appear to offer a more nuanced view of tumor and normal tissue-resident myeloid lineage cells, in comparison to T cells, in terms of continuity vs.\ separation of cell states. 
Unlike T cells, which primarily displayed continuous activation transitions, we observed sharper state delineations in myeloid populations. 
This difference between T cells and myeloid cells was likely due to a markedly less appreciated developmentally established myeloid cell heterogeneity, whose understanding has started to emerge only recently (Perdiguero and Geissmann, 2016). % citations 
Indeed, the phenotypic expansion in myeloid cells was associated with activation of macrophages and monocytes and emergence of pDC subsets distinct from cDCs. 
However, our analyses also showed common features to those in T cells, including gene expression covariance identifying cell clusters, and an expansion of immune phenotypic space in breast tumor as compared to normal breast tissue.

Similarly to T cells, we have not observed discrete states of myeloid cell activation/differentiation such as M1 or M2 macrophages or myeloid derived suppressor cells. 
In contrast, we found both M1 and M2 associated genes frequently expressed in the same cells, positively correlated with one another and following the same activation trajectory. 
Furthermore, we found that covariance patterns between gene markers associated with the M1 and M2 model show rich diversity, and help distinguish the three TAM clusters. 
These results challenge the prevailing model of macrophage activation, wherein M1 and M2 activation states either exist as mutually exclusive discrete states or macrophages reside along a spectrum between the two states with a negatively correlated expression of M1 and M2-associated genes. 
Our findings solidify and reinforce previous reports from the bulk analysis of tumor-associated macrophages in mouse models of oncogene-driven breast cancer and analysis of myeloid cells in lung and kidney cancer using mass-cytometry (Chevrier et al., 2017; Franklin et al., 2014; Lavin et al., 2017). % citation
Notably, we observed more patient-specific variation in myeloid lineage cells than in T cells, with the frequency of the former ranging from just over 10\% to over 50\% in individual patients.  
Individual clusters similarly exhibited ranges of patient specificity.
The large patient effect in myeloid cells suggests that attempts at generalized targeting or reprogramming of suppressive myeloid cell populations are not likely to yield uniform responses and personalization at the patient level may be needed.

Thus, our findings show that studying average gene expression across groups of cells fails to characterize heterogeneity in co-expression of genes, and by extension their potential suitability as therapeutic co-targets. 
Single-cell RNA-sequencing analyzed using Biscuit, as shown here, allows for inference of accurate and meaningful covariance parameters --- indeed, the algorithm takes into consideration these covariance values when defining clusters. 
This makes it possible to query in a precise manner how numerous functionally and therapeutically important immune markers are co-expressed at the level that matters: that of individual cells. 
Our characterization of the immune cell subsets inhabiting primary solid tumor and the corresponding normal tissue, and their heterogeneity within a given patient and between different patients revealed expansions of a continuous ``phenotypic space'' as a principal feature of the two main cellular targets of cancer immunotherapy - T cells and myeloid cells. 
These observations, along with the resulting extensive immune single- cell RNA-seq datasets and the comprehensive analytical platform, will facilitate better knowledge understanding of potential mechanisms behind immune cell contributions to promoting and opposing tumor progression. 

\section{Validation \& Follow-up Experimentation}
The logical follow-up of this atlasing project is to confirm that the results observed in our experiments exclusively result from biological signals and not technical aspects of scRNA-seq. 
Our observations at the cellular level of the correspondence of the FACS-sorted populations to the observations in scRNA-seq are suggestive that our analyses mirror the biology represented in the tissues, albeit with reduced representation of small-volume cells. 
While not all studies carry out independent validations, additional confirmations of the intra-cellular covariance observations are warranted. % citation 
Previous studies \citep{Shalek2013,Shalek2014} have focused on microsocopy-based confirmations, using either single-molecule fluorescence in-situ hybridization (FISH), immuno-histochemistry, or immuno-fluorescence approaches to measure RNA or protein abundances. 
These approaches are problematic to apply to our data. 
Unlike many previous studies, we are working with rare immune isolates from complex patient tissues, rather than tumors, cell lines, or mouse models, each of which contain plentiful cells of the type under study.
Our most interesting hypotheses, about T Regulatory cell co-expression, also occur in the rares cell types under study (Figure~\ref{fig:s1d}). These problems combine to make it un-economical to detect an adequate number of cells through microscopy-based approaches. 
Instead, since the majority of our findings center on surface proteins, we believe that FACS sorting of T Regulatory cells based upon CTLA-4, TIGIT, and GITR, the markers with differential covariance, would be a suitable approach to confirm the functional nature of our discoveries. 
As such, we are in the process of profiling several additional patients with FACS, and expect to observe patient-specific co-expression differences: while all patients are expected to display T Regulatory cells with CTLA-4, TIGIT, and GITR, we expect that co-variance will differ across patients.  

A second question raised by this experiment is the cause of the observed T cell diversity. 
If the observed diversity is indeed correlated with TCR repertoire, the phenotypic volume of T cells should increase with TCR diversity. 
Within weeks of the completion of the studies described in this dissertation, 10x genomics released a kit that allows the simultaneous profiling of the 5' transcriptome and the TCR\@.
We will apply this kit to the FACS-sorted cells from the above patients, and will be interested to see if the TCR diversity correlates with phenotypic volume, or if our observations were indeed instead driven by the diversity of signaling molecules and gradients present within the tumor microenvironment. 
In other experiments, we are characterizing two TCRs with known epitope targets from a mouse cancer model using scRNA-seq\. these experiments should provide convergent evidence: if significant differences in expression are observed between the epitope-specific TCR clones, this will support the idea that TCR functionality may be responsible for siginficant variation in T cell mRNA expression profiles. 

Taken holistically, these experiments highlight the incredible diversity of tumor microenvironments; tumors bombard immune cells with diverse panels of cytokines, chemokines, and growth factors. 
Our observations strongly suggest that combinations of these stimuli are responsible for the diversity of observed phenotypic profiles in TILs. 
Thus these results suggest that prediction of phenotypic profiles and checkpoint expression, and therefore druggability of T Regulatory or t-effector cells, depends extensively upon characterization of T cell responses to complex cocktails of stimuli. 
The diversity of cancers would require profiling of a very large number of tumors, and may thus place this predictive goal out of reach for some time. 
Nevertheless, it may still be possible to profile target cells, and by observing their checkpoint expression, enable immune functionality despite not understanding how the cells came to express the particular set of checkpoint markers they are presenting.
Because droplet-based scRNA-seq requires cells be dissociated to flow through the encapsulation devices, these experiments cannot directly observe the microenvironment of the individual cells that we infer to contribute significantly to their phenotypes. 
However, an advantage of scRNA-seq is that we are able to measure many thousands of cells per experiment, and as such, if environmental characteristics of a cell could be quantified, and cells could be stratified in terms of similar environments, it would create an immediate and powerful method through which to quantify the effect of tissue microenvironment on each cell type under study. 
Perhaps more critically, while we were able to confirm the presence of significant immune infiltrate in each of our patients using immuno-histochemistry, we cannot distinguish cells that were directly in contact with tumor cells from those restricted to the periphery of a tumor or those in contact only with stromal or other immune cells. 
If differences in the phenotypes of these different spatial contexts and the abundance of these cells could be posed as independent variables, it would be of great interest to correlate with drug response outcomes in clinical trials.  
There are emerging approaches capable of spatially profiling cells' transcriptomes and proteomes, and I believe that large-scale application of these technologies to patient tumors is both a logical next step, and an exciting opportunity to make rapid progress in understanding cancer immunology after controlling for these factors. 

\section{Implications of the SEQC Framework and Future Directions}
% what did SEQC accomplish and is it accomplishing
% how to improve upon SEQC
In addition to providing the basis for this immune atlas, SEQC also served as the processing framework for several other published studies. 
Its filters were used to generate interpretable data in a Fluidigm C1-based study wherein cells were cultured inside the C1 device, enabling single-cell live imaging of NF-$\kappa$B activation dynamics following LPS stimulation \citep{Lane2017}. 
SEQC was also used to generate the data that was used to develop the MAGIC \citep{Dijk2017} BISCUIT \citep{Prabhakaran2016,Azizi2017} and Wishbone \citep{Setty2016} algorithms. 
Early iterations of the processing methods were used to generate microplate-based single-cell sequencing data, and it is used in active production by Memorial Sloan Kettering Cancer Center to process the institute's droplet-based InDrop, 10x, and Nucleus-sequencing data. 
In addition, it has been used in at least one instance to process Drop-seq, 4 iterations of InDrop chemistry, and 3 versions of Mars-seq chemistry. 
In total, SEQC has processed over 250 datasets from diverse chemistries, tissues and multiple organisms.  
Finally, SEQC's demonstrated modularity and flexibility provoked the Human Cell Atlas to adopt the SEQC framework as the first draft of the 3' analysis pipeline that it will use in the secondary analysis of droplet-based sequencing data. 
The Human Cell Atlas project is expected to generate data at petabyte scale, a task made simple by SEQC's ability to scale using the Cloud.  

The development of SEQC, and its ability to deliver clean cellular phenotypes in spite of significant experimental noise, was instrumental in enabling he constribution of the tumor atlas. 
However, it is important to remember that the atlas would not have been possible at all if not for droplet-based sequencing.
Underlying the data generated in this dissertation were 2 major barcode redesigns, made possible in part by analysis metrics generated by InDrop providing clear feedback to experimenters about what aspects of the technology required improvement. 
The Pe'er lab has since iterated through $3$ major changes to the InDrop chemistry, and we expect more will follow.  

Before the SEQC framework was mature, the ability to process data would often lag weeks behind the development of new library construction approaches, significantly retarding our ability to make adjustments that would facilitate analysis of primary immune cells. 
Initially, the technologies were iterating much more quickly than the computational approaches. With the mature SEQC pipeline, we are now normally able to iterate and produce appropriate computational approaches for changes in library construction within a few weeks.  
In cases where appropriate algorithms exist, we can often make necessary changes in hours. 

Also symptomatic of this problem, novel technologies are often published with data processing tools that are inadequate, or at least sub-optimal. 
For example, sNuc-Seq \citep{Habib2016} and DroNc-seq \citep{Habib2017} are exciting methods that sequence nuclei instead of complete cells.
This difference promises to enlarge the scope of samples that can be processed with scRNAsq, since nuclear membranes are more robust than cell membranes, frozen sample or samples with more degradation can be sequenced with these methods. 

However, the data generated by these approaches were processed using tools designed for full-cell sequencing, ignoring the fact that many RNA in nuclei are found in pre-spliced ``pre-mRNA'' form and would therefore contain intronic reads, which are discarded by full-cell analysis pipelines. 
This introduces a data loss of 15-30\%, and strongly indicates that the technologies are still iterating much more rapidly than the computational methods. 
This highlights that in addition to a pipeline for processing data, the field would benefit from a framework for rapidly mixing and matching algorithms that consume and produce standard sequencing data types and file formats.  

In order to improve computational iteration speed, SEQC and frameworks like it must be made more portable and trivial to use. 
Because different labs and institutions have back-end compute server frameworks that can be idiosyncratic and utilize queueing architectures that are often incompatible with one another, this is currently difficult to achieve. 
While SEQC has unprecedented flexibility, it is still limited to being run on single physical or virtual machines, and can only easily be run on laptops, desktops, local compute servers or on Amazon web services. 
There are technological advances that are being developed which will solve these problems, and improve method portability not just for processing frameworks like SEQC, but also increase the portability of analysis algorithms like BISCUIT, MAGIC, and Wishbone.  

The first of these technologies are container services like Docker. 
Docker is a relatively mature scripting framework for constructing a lightweight software image that contains all of the installed dependencies for an particular software package, including the OS layer, and the software package itself. 
Docker images support versioning, meaning that a properly constructed docker image should go on working forever, regardless of what changes may happen to the greater ecosystem of the programming language or advancements in packages it depends upon. 
Docker is used extensively by companies to package software to make it more robust, reliable, and easier to ship to customers. 

However, this approach has not made much of an inroad in academia, perhaps because of a higher focus on publication than eventual usability. 
SEQC was initially programmed to follow a strategy similar to docker, but with some limitations.
We first programmed an Amazon Machine Image, which is like an Amazon-specific docker container. 
As a result, it only worked on Amazon's Elastic Cloud, and we needed to use separate installation instructions for local installations. 

Docker, by contrast, can run on Amazon, but also on Google cloud, and on mac and PC.
For these reasons, SEQC has been made available in a docker container, which can be pulled from the dockerhub repository at ambrosejcarr/seqc:1.0.0.
As a result, SEQC can now be used on any operating system, without needing expert knowledge of their operating system to install any dependencies. 

Because containers allow a developer to easily produce ready-to-run container, if methods consume and produce standard data types, such as BAM and FASTQ formats, then multiple methods that accomplish the same task can be benchmarked and interchanged, and these pieces can be woven together through workflow languages. 
Two workflow languages, Common Workflow Language (CWL) and Workflow Design Language (WDL) are being developed to serve this purpose, and these frameworks have been adopted by the Broad Institute, University of Santa Cruz, European Bioinformatics Institute, Chan Zuckerberg Initiative, and Human Cell Atlas, among others, to serve as a framework for building open software that can benefit the community. 

A framework based upon SEQC is presently being written in WDL, and once complete, its computational approaches will serve as a standard against which alternative algorithms may be benchmarked.
Through creative use of control datasets, I hope to discover the optimal combination of computational approaches that produce the highest quality sequencing data for, at first, 3' droplet-based sequencing approaches. 
Nucleus sequencing, 10x Genomics, and other approaches all utilize a common set of core methods, and as such, the InDrop backbone can be adapted to suit other technologies. 
It is my hope that by refining this pipeline and pairing it with any number of front-end suites of analysis tools, we can speed up the experiment, analyze, refine loop for advances in single cell sequencing, enabling faster technological development, opening doors to generate many more cell atlases like the one described here, each with clear clinical implications. 
